File: frontend/src/utils/textCleaner.ts
/**
 * Text Cleaning Utilities
 * Strips markdown, emoji, and special characters for clean TTS and whiteboard display.
 */

/** Clean text for TTS speech — removes all formatting artifacts */
export function cleanForTTS(text: string): string {
  if (!text) return ''

  let s = text

  // Remove code blocks
  s = s.replace(/```[\s\S]*?```/g, ' code block ')
  s = s.replace(/`[^`]+`/g, '')

  // Remove markdown headers
  s = s.replace(/#{1,6}\s*/g, '')

  // Remove bold/italic markers
  s = s.replace(/\*{1,3}/g, '')
  s = s.replace(/_{1,3}/g, '')

  // Remove markdown links [text](url) -> text
  s = s.replace(/\[([^\]]*)\]\([^)]*\)/g, '$1')

  // Remove bare URLs
  s = s.replace(/https?:\/\/\S+/g, 'link')

  // Remove arrows and special symbols
  s = s.replace(/→/g, 'becomes')
  s = s.replace(/->/g, 'becomes')
  s = s.replace(/=>/g, 'becomes')
  s = s.replace(/---+/g, '')
  s = s.replace(/—/g, ', ')
  s = s.replace(/–/g, ', ')

  // Remove emoji (common ranges)
  s = s.replace(
    /[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2702}-\u{27B0}\u{24C2}-\u{1F251}\u{1F900}-\u{1F9FF}\u{2600}-\u{26FF}\u{FE00}-\u{FE0F}\u{200D}\u{2B50}\u{2B55}]+/gu,
    ' '
  )

  // Remove list markers at start of lines
  s = s.replace(/^\s*[-*+]\s+/gm, '')
  s = s.replace(/^\s*\d+\.\s+/gm, '')

  // Collapse whitespace
  s = s.replace(/\n{2,}/g, '. ')
  s = s.replace(/\n/g, ' ')
  s = s.replace(/\s{2,}/g, ' ')

  return s.trim()
}

/** Clean text for whiteboard display — keeps structure, removes markdown noise */
export function cleanForWhiteboard(text: string): string {
  if (!text) return ''

  let s = text

  // Remove markdown bold/italic
  s = s.replace(/\*{1,3}/g, '')
  s = s.replace(/_{1,3}/g, '')

  // Remove headers
  s = s.replace(/#{1,6}\s*/g, '')

  // Clean links
  s = s.replace(/\[([^\]]*)\]\([^)]*\)/g, '$1')

  // Remove code fences
  s = s.replace(/```\w*\n?/g, '')
  s = s.replace(/`/g, '')

  // Remove horizontal rules
  s = s.replace(/^---+\s*$/gm, '')

  // Remove emoji
  s = s.replace(
    /[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2702}-\u{27B0}\u{24C2}-\u{1F251}\u{1F900}-\u{1F9FF}\u{2600}-\u{26FF}\u{FE00}-\u{FE0F}\u{200D}\u{2B50}\u{2B55}]+/gu,
    ''
  )

  // Collapse excess whitespace
  s = s.replace(/\n{3,}/g, '\n\n')
  s = s.replace(/[ \t]{2,}/g, ' ')

  return s.trim()
}

/** Clean text for subtitle display — single line, no symbols */
export function cleanForSubtitle(text: string): string {
  let s = cleanForTTS(text)
  s = s.replace(/[\n\r]+/g, ' ')
  s = s.replace(/\.+$/, '')
  return s.trim()
}

File: backend/app/utils/text_cleaner.py
"""
Text Cleaning Utilities — Strip markdown, emoji, and special characters
for clean TTS speech and whiteboard display.
"""
import re


def clean_for_tts(text: str) -> str:
    """Clean text for text-to-speech output. Removes all formatting artifacts."""
    if not text:
        return ""

    s = text

    # Remove code blocks entirely (replace with "code block")
    s = re.sub(r'```[\s\S]*?```', ' code block ', s)
    s = re.sub(r'`[^`]+`', '', s)

    # Remove markdown headers
    s = re.sub(r'#{1,6}\s*', '', s)

    # Remove bold/italic markers
    s = re.sub(r'\*{1,3}', '', s)
    s = re.sub(r'_{1,3}', '', s)

    # Remove markdown links [text](url) -> text
    s = re.sub(r'\[([^\]]*)\]\([^)]*\)', r'\1', s)

    # Remove bare URLs
    s = re.sub(r'https?://\S+', 'link', s)

    # Remove arrows and special symbols
    s = s.replace('→', 'becomes')
    s = s.replace('->', 'becomes')
    s = s.replace('=>', 'becomes')
    s = s.replace('---', '')
    s = s.replace('—', ', ')
    s = s.replace('–', ', ')

    # Remove emoji (Unicode emoji ranges)
    s = re.sub(
        r'[\U0001F600-\U0001F64F'   # emoticons
        r'\U0001F300-\U0001F5FF'     # symbols & pictographs
        r'\U0001F680-\U0001F6FF'     # transport & map
        r'\U0001F1E0-\U0001F1FF'     # flags
        r'\U00002702-\U000027B0'     # dingbats
        r'\U000024C2-\U0001F251'     # misc
        r'\U0001f900-\U0001f9FF'     # supplemental
        r'\U00002600-\U000026FF'     # misc symbols
        r'\U0000FE00-\U0000FE0F'     # variation selectors
        r'\U0000200D'                # ZWJ
        r'\U00002B50\U00002B55'      # stars
        r']+',
        ' ',
        s,
        flags=re.UNICODE,
    )

    # Remove list markers at start of lines
    s = re.sub(r'^\s*[-*+]\s+', '', s, flags=re.MULTILINE)
    s = re.sub(r'^\s*\d+\.\s+', '', s, flags=re.MULTILINE)

    # Collapse whitespace
    s = re.sub(r'\n{2,}', '. ', s)
    s = re.sub(r'\n', ' ', s)
    s = re.sub(r'\s{2,}', ' ', s)

    return s.strip()


def clean_for_whiteboard(text: str) -> str:
    """Clean text for whiteboard display. Keeps structure but removes markdown noise."""
    if not text:
        return ""

    s = text

    # Remove markdown bold/italic
    s = re.sub(r'\*{1,3}', '', s)
    s = re.sub(r'_{1,3}', '', s)

    # Remove markdown headers
    s = re.sub(r'#{1,6}\s*', '', s)

    # Clean links
    s = re.sub(r'\[([^\]]*)\]\([^)]*\)', r'\1', s)

    # Remove code fences
    s = re.sub(r'```\w*\n?', '', s)
    s = re.sub(r'`', '', s)

    # Remove horizontal rules
    s = re.sub(r'^---+\s*$', '', s, flags=re.MULTILINE)

    # Remove emoji
    s = re.sub(
        r'[\U0001F600-\U0001F64F'
        r'\U0001F300-\U0001F5FF'
        r'\U0001F680-\U0001F6FF'
        r'\U0001F1E0-\U0001F1FF'
        r'\U00002702-\U000027B0'
        r'\U000024C2-\U0001F251'
        r'\U0001f900-\U0001f9FF'
        r'\U00002600-\U000026FF'
        r'\U0000FE00-\U0000FE0F'
        r'\U0000200D'
        r'\U00002B50\U00002B55'
        r']+',
        '',
        s,
        flags=re.UNICODE,
    )

    # Collapse excess whitespace but keep single newlines
    s = re.sub(r'\n{3,}', '\n\n', s)
    s = re.sub(r'[ \t]{2,}', ' ', s)

    return s.strip()


def clean_for_subtitle(text: str) -> str:
    """Clean text for subtitle overlay. Single line, no symbols."""
    s = clean_for_tts(text)
    # Ensure single line
    s = re.sub(r'[\n\r]+', ' ', s)
    # Remove trailing periods for subtitle feel
    s = s.rstrip('.')
    return s.strip()